site/_posts/2021-09-16-how-to-authenticate-your-react-app.md

```diff
@@ -0,0 +1,896 @@
+    request(
+      // POST request to /introspect endpoint
+      {
+        method: 'POST',
+        uri: `http://localhost:${config.fusionAuthPort}/oauth2/introspect`,
+        form: {
+          'client_id': config.clientID,
+          'token': req.session.token
+        }
+      },
+
+      // callback
+      (error, response, body) => {
+        let introspectResponse = JSON.parse(body);
+
+        // valid token -> get more user data and send it back to the react app
+        if (introspectResponse.active) {
+          request(
+            // GET request to /registration endpoint
+            {
+              method: 'GET',
+              uri: `http://localhost:${config.fusionAuthPort}/api/user/registration/${introspectResponse.sub}/${config.applicationID}`,
+              json: true,
+              headers: {
+                'Authorization': config.apiKey
+              }
+            },
+
+            // callback
+            (error, response, body) => {
+              res.send(
+                {
+                  token: {
+                    ...introspectResponse,
+                  },
+                  ...body
+                }
+              );
+            }
+          );
+        }
+
+        // expired token -> send nothing
+        else {
+          req.session.destroy();
+          res.send({});
+        }
+      }
+    );
+  }
+
+  // no token -> send nothing
+  else {
+    res.send({});
+  }
+});
+
+module.exports = router;
+```
+
+First, the POST request to `/introspect`: this endpoint requires our Client ID and, of course, the Access Token we want to decode. Just like the token endpoint, this endpoint access form encoded data, so we are using the form option.
+
+In the callback from that request, we’ll parse the body into usable JSON and check that active value. If it’s good, we’ll go ahead and make the `/registration` request, which requires a User ID (“subscriber” or sub in OAuth jargon) and our Client ID—we need both, because one user can have different data in each FusionAuth application.
+
+Note that this step doesn’t actually require an Access Token. The data from FusionAuth’s `/api/user/registration` API can be retrieved for any user, including one that isn’t logged in. You could use this endpoint to show info on a user’s profile or other public page. This API in FusionAuth is not part of the OAuth specification but we added it to our example to show a more complete user data interaction. Also, this API requires that you pass in a FusionAuth API key in the Authorization header. Our example is pulling the FusionAuth API key from our global `config.js` file.
+
+Finally, in the registration callback, we’ll parse the body returned from `/registration` and `res.send` everything back to the React client.
+
+Now, go through the login process, and you should see “Welcome, [your email address]!”.
+
+That’s login sorted. The next thing you’ll probably want to tackle is logout.
+
+## Logging Out
```

Comment:
Title should be `Logging out`. For blog posts we don't capitalize every word, just proper nouns and the first one.